zio-http/src/main/scala/zhttp/endpoint/CanConstruct.scala

```diff
@@ -23,11 +23,12 @@ object CanConstruct {
     override type EOut = E
 
     override def make(route: Endpoint[A], f: Request.ParameterizedRequest[A] => Response[R, E]): HttpApp[R, E] =
-      Http.collect[Request] { case req =>
-        route.extract(req) match {
-          case Some(value) => f(Request.ParameterizedRequest(req, value))
-          case None        => Response.fromHttpError(HttpError.NotFound(req.url.path))
-        }
+      Http.collect[Request] {
+        case req if route.extract(req).isDefined =>
+          route.extract(req) match {
```

Comment:
Prevent extraction from happening twice

Reply:
```scala
Http.collect[Request] { case req =>
          route.extract(req) match {
            case Some(value) => Http.succeed(f(Request.ParameterizedRequest(req, value)))
            case None        => Http.empty
          }
        }
        .flatten
```
Something like this might solve the issue. Don't know which one performs better jmh benchmark might help here.
@tusharmath  Suggestion